### PR TITLE
chore(w3c): add last parent id to tracestate

### PR DESCRIPTION
**Pull request by mabdinur**
With this change span_id will be encoded in the w3c tracestate header. Propagating span_id in both tracestate and traceparent will allow us to detect non Datadog spans in distributed traces. These span ids will be sent to trace intake services via the dd.parent_id tag. 

## Checklist

- [x] Change(s) are motivated and described in the PR description
- [x] Testing strategy is described if automated tests are not included in the PR
- [x] Risks are described (performance impact, potential for breakage, maintainability)
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed or label `changelog/no-changelog` is set
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/))
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))
- [x] If this PR changes the public interface, I've notified `@DataDog/apm-tees`.
- [x] If change touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.

## Reviewer Checklist

- [x] Title is accurate
- [x] All changes are related to the pull request's stated goal
- [x] Description motivates each change
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [x] Testing strategy adequately addresses listed risks
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] Release note makes sense to a user of the library
- [x] Author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/propagation/http.py [hunk 1]
```diff
@@ -792,8 +803,9 @@ def _get_context(trace_id, span_id, trace_flag, ts, meta=None):
                     tracestate_values = None
 
                 if tracestate_values:
-                    sampling_priority_ts, other_propagated_tags, origin = tracestate_values
+                    sampling_priority_ts, other_propagated_tags, origin, lpid = tracestate_values
                     meta.update(other_propagated_tags.items())
+                    meta["_dd.lp.id"] = str(lpid)
```
Comment (mabdinur) [inline ddtrace/propagation/http.py:808]:
```suggestion
                    if lpid is not None:
                        meta["_dd.lp.id"] = str(lpid)
```

**Comment by mabdinur** [comment]
> ## Benchmarks
> Benchmark execution time: 2024-02-09 22:40:38
> 
> Comparing candidate commit [ac42901](https://github.com/DataDog/dd-trace-py/commit/ac42901b59f5dd083cfca27d7886c62817595a25) in PR branch `munir/w3c-phase2-tracestate` with baseline commit [4fd076a](https://github.com/DataDog/dd-trace-py/commit/4fd076ac83da668abf2e931dc1fd846f7ce66065) in branch `main`.
> 
> Found 1 performance improvements and 8 performance regressions! Performance is the same for 186 metrics, 9 unstable metrics.
> 
> #### scenario:flasksimple-appsec-telemetry
> * 🟩 `execution_time` [-258.684µs; -213.024µs] or [-3.959%; -3.260%]
> 
> #### scenario:httppropagationinject-ids_only
> * 🟥 `execution_time` [+708.047ns; +762.440ns] or [+8.965%; +9.653%]
> 
> #### scenario:httppropagationinject-with_all
> * 🟥 `execution_time` [+1.110µs; +1.196µs] or [+5.951%; +6.410%]
> 
> #### scenario:httppropagationinject-with_dd_origin
> * 🟥 `execution_time` [+934.940ns; +990.797ns] or [+7.659%; +8.117%]
> 
> #### scenario:httppropagationinject-with_priority_and_origin
> * 🟥 `execution_time` [+840.435ns; +924.542ns] or [+6.027%; +6.630%]
> 
> #### scenario:httppropagationinject-with_sampling_priority
> * 🟥 `execution_time` [+846.009ns; +889.623ns] or [+8.578%; +9.021%]
> 
> #### scenario:httppropagationinject-with_tags
> * 🟥 `execution_time` [+1.015µs; +1.096µs] or [+7.523%; +8.127%]
> 
> #### scenario:httppropagationinject-with_tags_invalid
> * 🟥 `execution_time` [+806.416ns; +890.681ns] or [+5.169%; +5.709%]
> 
> #### scenario:httppropagationinject-with_tags_max_size
> * 🟥 `execution_time` [+1.146µs; +1.231µs] or [+8.207%; +8.816%]

This overhead appears unavoidable.

File: ddtrace/propagation/http.py [hunk 2]
```diff
@@ -815,8 +821,13 @@ def _inject(span_context, headers):
             headers[_HTTP_HEADER_TRACEPARENT] = tp
             # only inject tracestate if traceparent injected: https://www.w3.org/TR/trace-context/#tracestate-header
             ts = span_context._tracestate
-            if ts:
-                headers[_HTTP_HEADER_TRACESTATE] = ts
+            # Adds last datadog parent_id to tracestate. This tag is used to reconnect a traces with missing spans
```
Comment (erikayasuda) [inline ddtrace/propagation/http.py:824]: nit:
```suggestion
            # Adds last datadog parent_id to tracestate. This tag is used to reconnect a trace with missing spans
```

File: ddtrace/propagation/http.py [hunk 3]
```diff
@@ -815,8 +821,13 @@ def _inject(span_context, headers):
...
+            if "dd=" in ts:
+                ts = ts.replace("dd=", "dd=p:{:016x};".format(span_context.span_id or 0))
+            else:
+                ts = "dd=p:{:016x}".format(span_context.span_id or 0)
```
Comment (erikayasuda) [inline ddtrace/propagation/http.py:828]: Is this just accounting for cases where there are multiple occurrences of `dd=` in the if case, and empty/malformed tracestates in the else case?

**Review by bouwkast — APPROVED** [review]
Looks good to me based on the doc! Thanks @mabdinur

File: ddtrace/propagation/http.py (hunk 3 above)
Comment (mabdinur) [reply ddtrace/propagation/http.py:828]: Having the else clause allows us to account for the case where `dd=` is not in the tracestate but this case will only occur if there is a bug in our tracer (spans do not have sampling decsions or tracestate is malformed).

**Pull request merged**